atags/atags.1
.\" Copyright (c) 1987, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ctags.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: release/10.0.0/usr.bin/ctags/ctags.1 216370 2010-12-11 08:32:16Z joel $
.\"
.Dd December 31. 2018
.Dt ATAGS 1
.Os "Orthanc Systems"
.Sh NAME
.Nm atags
.Nd create an
acme
tags
file
.Sh SYNOPSIS
.Nm
.Op Fl dqtv
.Op Fl f Ar tagsfile
.Ar
.Sh DESCRIPTION
The
.Nm
utility makes a
tags
file,
suitable for
.Xr acme 1
and the Plan\ 9 plumber,
from the specified C,
Fortran,
lex,
Lisp,
Pascal,
and yacc sources.
The tags file indexes the location of objects in a group of source code
files.
Each line of the tags file contains the object name,
along with the file and line number
at which it
is defined.
Using the
tags
file,
.Sy acme
can quickly locate these object definitions.
Depending upon the options provided to
.Nm ,
objects will consist of subroutines, typedefs, defines, structs,
enums and unions.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl d
Create tags for
.Li #defines
that do not take arguments;
.Li #defines
that take arguments are tagged automatically.
.It Fl f
Place the tag descriptions in a file called
.Ar tagsfile .
The default behaviour is to place them in a file called
.Pa ATAGS .
.It Fl q
Suppress warning diagnostics.
.It Fl t
Create tags for typedefs, structs, unions, and enums.
.It Fl v
An index of the form expected by
.Xr vgrind 1
is produced on the standard output.
This listing
contains the object name, file name, and page number (assuming 64
line pages).
Since the output will be sorted into lexicographic order,
it may be desired to run the output through
.Xr sort 1 .
E.g.:
.Bd -literal -offset indent
atags -v files | sort -f > index
vgrind -x index
.Ed
.El
.Pp
Files whose names end in
.Pa .c
or
.Pa .h
are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in
.Pa .y
are assumed to be
.Xr yacc 1
source files.
Files whose names end in
.Pa .l
are assumed to be Lisp files if their
first non-blank character is
.Ql \&; ,
.Ql \&( ,
or
.Ql \&[ ,
otherwise, they are
treated as
.Xr lex 1
files.
Other files are first examined to see if they
contain any Pascal or Fortran routine definitions, and, if not, are
searched for C style definitions.
.Pp
The tag
.Dq Li main
is treated specially in C programs.
The tag formed
is created by prepending
.Ql M
to the name of the file, with the
trailing
.Pa .c
and any leading pathname components removed.
This makes use of
.Nm
practical in directories with more than one
program.
.Pp
The
.Xr yacc 1
and
.Xr lex 1
files each have a special tag.
.Dq Li yyparse
is the start
of the second section of the
.Xr yacc 1
file, and
.Dq Li yylex
is the start of
the second section of the
.Xr lex 1
file.
.Sh FILES
.Bl -tag -compact
.It Pa ATAGS
The default tags file
.El
.Sh EXIT STATUS
The
.Nm
utility exits with a value of 1 if an error occurred, 0 otherwise.
Duplicate objects are not considered errors.
.Sh SEE ALSO
.Xr acme 1 ,
.Xr ctags 1
.Pp
.Sh BUGS
Recognition of functions, subroutines and procedures
for Fortran and Pascal is done in a very simpleminded way.
No attempt
is made to deal with block structure; if you have two Pascal procedures
in different blocks with the same name you lose.
The
.Nm
utility does not
understand about Pascal types.
.Pp
The method of deciding whether to look for C, Pascal or
Fortran
functions is a hack.
.Pp
The
.Nm
utility relies on the input being well formed, and any syntactical
errors will completely confuse it.
It also finds some legal syntax
confusing; for example, since it does not understand
.Li #ifdef Ns 's
(incidentally, that is a feature, not a bug), any code with unbalanced
braces inside
.Li #ifdef Ns 's
will cause it to become somewhat disoriented.
In a similar fashion, multiple line changes within a definition will
cause it to enter the last line of the object, rather than the first, as
the searching pattern.
The last line of multiple line
.Li typedef Ns 's
will similarly be noted.


atags/atags.c
/*
 * Copyright (c) 1987, 1993, 1994, 1995
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1987, 1993, 1994, 1995\n\
	The Regents of the University of California.  All rights reserved.\n";
static char sccsid[] = "@(#)ctags.c	8.4 (Berkeley) 2/7/95";
static char fbsdid[] = "$FreeBSD: release/10.0.0/usr.bin/ctags/ctags.c 216370 2010-12-11 08:32:16Z joel $";
#endif

#include <err.h>
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "atags.h"

/*
 * ctags: create a tags file
 */

NODE	*head;			/* head of the sorted binary tree */

				/* boolean "func" (see init()) */
bool	_wht[256], _etk[256], _itk[256], _btk[256], _gd[256];

FILE	*inf;			/* ioptr for current input file */
FILE	*outf;			/* ioptr for tags file */

long	lineftell;		/* ftell after getc( inf ) == '\n' */

int	lineno;			/* line number of current line */
int	dflag;			/* -d: non-macro defines */
int	qflag;			/* -q: suppress warnings */
int	tflag;			/* -t: create tags for typedefs */
int	vflag;			/* -v: vgrind style index output */

char	*curfile;		/* current input file name */
char	lbuf[LINE_MAX];

void	init(void);
void	find_entries(char *);
static void usage(void);

int
main(int argc, char **argv)
{
	const char	*outfile = "ATAGS";	/* output file */
	int		exit_val;		/* exit value */
	int		step;			/* step through args */
	int		ch;			/* getopts char */

	setlocale(LC_ALL, "C");

	dflag = qflag = tflag = vflag = 0;
	while ((ch = getopt(argc, argv, "df:qtv")) != -1)
		switch(ch) {
		case 'd':
			dflag++;
			break;
		case 'f':
			outfile = optarg;
			break;
		case 'q':
			qflag++;
			break;
		case 't':
			tflag++;
			break;
		case 'v':
			vflag++;
			break;
		case '?':
		default:
			usage();
		}
	argv += optind;
	argc -= optind;
	if (!argc)
		usage();

	setlocale(LC_COLLATE, "C");

	init();

	for (exit_val = step = 0; step < argc; ++step)
		if (!(inf = fopen(argv[step], "r"))) {
			warn("%s", argv[step]);
			exit_val = 1;
		}
		else {
			curfile = argv[step];
			find_entries(argv[step]);
			(void)fclose(inf);
		}

	if (head) {
		if (!(outf = fopen(outfile, "w")))
			err(1, "%s", outfile);
		put_entries(head);
		(void)fclose(outf);
	}
	exit(exit_val);
}

static void
usage(void)
{
	(void)fprintf(stderr, "usage: ctags [-dqtv] [-f tagsfile] file ...\n");
	exit(1);
}

/*
 * init --
 *	this routine sets up the boolean pseudo-functions which work by
 *	setting boolean flags dependent upon the corresponding character.
 *	Every char which is NOT in that string is false with respect to
 *	the pseudo-function.  Therefore, all of the array "_wht" is NO
 *	by default and then the elements subscripted by the chars in
 *	CWHITE are set to YES.  Thus, "_wht" of a char is YES if it is in
 *	the string CWHITE, else NO.
 *
 *	XXX Not at all utf8 aware.
 */
void
init(void)
{
	int	i;
	char	*sp;

	for (i = 0; i < 256; i++) {
		_wht[i] = _etk[i] = _itk[i] = _btk[i] = NO;
		_gd[i] = YES;
	}

	/* white space chars */
	for (sp = " \f\t\n"; *sp; sp++) {
		_wht[*sp] = YES;
	}

	/* token ending chars */
	for (sp = " \t\n\"'#()[]{}=-+%*/&|^~!<>;,.:?"; *sp; sp++) {
		_etk[*sp] = YES;
	}

	/* valid in-token chars */
	for (sp = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz0123456789"; *sp; sp++) {
		_itk[*sp] = YES;
	}

	/* token starting chars */
	for (sp = "ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"; *sp; sp++) {
		_btk[*sp] = YES;
	}

	/* invalid after-function chars */
	for (sp = ",;"; *sp; sp++) {
		_gd[*sp] = NO;
	}
}

/*
 * find_entries --
 *	this routine opens the specified file and calls the function
 *	which searches the file.
 */
void
find_entries(char *file)
{
	char	*cp;

	lineno = 0;				/* should be 1 ?? KB */
	if ((cp = strrchr(file, '.'))) {
		if (cp[1] == 'l' && !cp[2]) {
			int	c;

			for (;;) {
				if (GETC(==, EOF))
					return;
				if (!iswhite(c)) {
					rewind(inf);
					break;
				}
			}
/* lisp */		if (strchr(";([", c)) {
				l_entries();
				return;
			}
/* lex */		else {
				/*
				 * we search all 3 parts of a lex file
				 * for C references.  This may be wrong.
				 */
				toss_yysec();
				(void)strcpy(lbuf, "%%$");
				pfnote("yylex", lineno);
				rewind(inf);
			}
		}
/* yacc */	else if (cp[1] == 'y' && !cp[2]) {
			/*
			 * we search only the 3rd part of a yacc file
			 * for C references.  This may be wrong.
			 */
			toss_yysec();
			(void)strcpy(lbuf, "%%$");
			pfnote("yyparse", lineno);
			y_entries();
		}
/* fortran */	else if ((cp[1] != 'c' && cp[1] != 'h') && !cp[2]) {
			if (PF_funcs())
				return;
			rewind(inf);
		}
	}
/* C */	c_entries();
}
